File: android/src/main/kotlin/com/jameskarl/amap/map/MapMessageHandler.kt
package com.jameskarl.amap.map

import android.util.Log
import com.amap.api.maps.TextureMapView
import com.jameskarl.amap.AmapPlugin
import com.jameskarl.amap.map.bean.toMarkerData
import com.jameskarl.shop.toJson
import io.flutter.plugin.common.BasicMessageChannel
import io.flutter.plugin.common.PluginRegistry
import io.flutter.plugin.common.StringCodec
import org.json.JSONObject

class MapMessageHandler(
        private val registrar: PluginRegistry.Registrar,
        private val mapView: TextureMapView,
        viewId: Int
) : BasicMessageChannel.MessageHandler<String> {

    private val mapMethodChannelName = "${AmapPlugin.mapViewType}/map$viewId"

    private val messageChannel: BasicMessageChannel<String> = BasicMessageChannel<String>(registrar.messenger(), mapMethodChannelName, StringCodec.INSTANCE)

    fun setup() {
        messageChannel.setMessageHandler(this)
        mapView.map.setOnMarkerClickListener {
            it.showInfoWindow()
            sendJsonMessageToFlutter(ReplyToFlutter.Success(MapMethods.onMarkerClicked, it.toMarkerData()))
            true
        }
        mapView.map.setOnMapClickListener {
            sendJsonMessageToFlutter(ReplyToFlutter.Success(MapMethods.onMapClicked, it))
        }
        mapView.map.setOnMapLoadedListener {
            Log.d("MAP", "onMapLoaded")
            sendJsonMessageToFlutter(ReplyToFlutter.Success(MapMethods.onMapLoaded))
        }
        Log.d("MAP", mapMethodChannelName)
    }

    private fun sendJsonMessageToFlutter(message: ReplyToFlutter) {
        val msg = message.toJson()
        messageChannel.send(msg)
    }

    override fun onMessage(message: String?, reply: BasicMessageChannel.Reply<String>) {
        if (message != null) {
            try {
                val json = JSONObject(message)
                val methodId = json["id"]
                val data = json.opt("data")
                if (methodId is String && methodId.isNotBlank()) {
                    MapMethods.handleMessage(mapView.map, methodId, data, reply)
                } else {
                    Log.e("MAP", "method id is absent.")
                    MapMethods.handleException(methodId?.toString()
                            ?: "???", "method id is absent", reply)
                }
            } catch (e: Throwable) {
                e.printStackTrace()
                MapMethods.handleException("???", e.localizedMessage, reply)
            }
        }
    }

}

File: ios/Classes/MapMessageHandler.swift
//
//  MapMessageHandler.swift
//  amap
//
//  Created by James Karl on 2019/9/11.
//

import Foundation
import MAMapKit

class MapMessageHandler : NSObject, MAMapViewDelegate {
    let registrar: FlutterPluginRegistrar
    let mapView: MAMapView
    let viewId: Int
    
    let mapMethodChannelName: String
    let messageChannel: FlutterBasicMessageChannel
    
    init(registrar: FlutterPluginRegistrar, mapView: MAMapView, viewId: Int) {
        self.registrar = registrar
        self.mapView = mapView
        self.viewId = viewId
        
        self.mapMethodChannelName = "\(SwiftAmapPlugin.mapViewType)/map\(viewId)"
        self.messageChannel = FlutterBasicMessageChannel.init(
            name: mapMethodChannelName,
            binaryMessenger:  registrar.messenger(),
            codec: FlutterStringCodec.sharedInstance()
        )
    }
    
    func setup() {
        messageChannel.setMessageHandler({ message, reply in
            print("message from flutter \(String(describing: message))")
            if let messageData  = message as? String {
                self.onReceiveFlutterMessage(message: messageData, reply: reply)
            } else {
                self.sendJsonMessageToFlutter(message: ReplyToFlutter.failed(message:"message must be a json string"))
            }
        })
        mapView.delegate = self
    }
    
    func onReceiveFlutterMessage(message: String, reply: FlutterReply) {
        if let messageData  = message.data(using: .utf8) {
            if let json = (try? JSONSerialization.jsonObject(with: messageData, options: [])) as? [String: Any?] {
                if let methodId = json["id"] as? String {
                    MapMethods.handleMessage(mapView: self.mapView, methodId: methodId, data: json["data"] as Any?, reply: reply)
                } else {
                    self.sendJsonMessageToFlutter(message: ReplyToFlutter.failed(message:"method id NOT found \(String(describing: message))"))
                }
            } else  {
                self.sendJsonMessageToFlutter(message: ReplyToFlutter.failed(message:"parse josn message \(String(describing: message)) failed)"))
            }
        } else {
            self.sendJsonMessageToFlutter(message: ReplyToFlutter.failed(message:"message  \(String(describing: message)) must be a json string"))
        }
    }
    
    func mapViewDidFinishLoadingMap(_ mapView: MAMapView!) {
        print("xxxxx map loaded")
        sendJsonMessageToFlutter(message: ReplyToFlutter.success(id: MapMethods.onMapLoaded))
    }
    
    func sendJsonMessageToFlutter(message: ReplyToFlutter) {
        if let json = message.toJson() {
            messageChannel.sendMessage(json)
        } else {
            print("sendJsonMessageToFlutter failed \(message)")
        }
    }

    func mapView(_ mapView: MAMapView!, didSingleTappedAt coordinate: CLLocationCoordinate2D) {
        sendJsonMessageToFlutter(message: ReplyToFlutter.success(id: MapMethods.onMapClicked, data: coordinate.toLatLngData()))
    }
    
    func mapView(_ mapView: MAMapView!, didAnnotationViewTapped view: MAAnnotationView!) {
        sendJsonMessageToFlutter(message: ReplyToFlutter.success(id: MapMethods.onMarkerClicked, data: view.toAnnotationViewData()))
    }
    
}
